### PR TITLE
Vectorize preprocess for `detect/obb` loss calculation

### PR DESCRIPTION
**Pull request by Laughing-q**
inspired by https://github.com/ultralytics/ultralytics/pull/23937 by author @ahmet-f-gumustas

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
⚡ This PR speeds up target preprocessing in `loss.py` by replacing per-image Python loops with vectorized PyTorch indexing, making training more efficient without changing model behavior.

### 📊 Key Changes
- 🚀 Reworked two `preprocess()` paths in `ultralytics/utils/loss.py` to avoid looping over each image in a batch.
- 🔁 Replaced repeated `for j in range(batch_size)` and boolean masking logic with batched tensor indexing using:
  - `batch_idx`
  - cumulative `offsets`
  - `within_idx`
- 🧠 Converted image indices to `long` explicitly, improving indexing safety and consistency in PyTorch.
- 📦 For the second preprocessing block, targets are now copied into `packed_targets` first, then bounding boxes are scaled in one batched step before assignment.
- ✅ Output structure and downstream behavior remain the same: targets are still packed into fixed batch-aligned tensors for loss computation.

### 🎯 Purpose & Impact
- ⚡ **Faster training step preparation:** Vectorized tensor operations are typically much faster than Python-side loops, especially on GPUs.
- 📈 **Better scalability:** Helps when training with larger batch sizes or datasets containing many labels per image.
- 🧹 **Cleaner implementation:** The new logic is more optimized and better aligned with PyTorch best practices for batched operations.
- 🔒 **No expected user-facing changes:** Training results and APIs should stay the same, but preprocessing inside the loss pipeline should run more efficiently.
- 💻 **Potential practical benefit:** Users training YOLO models may see modest performance improvements and reduced overhead during training workloads.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Laughing-q, thank you for submitting this `ultralytics/ultralytics` 🚀 PR for vectorized `detect`/OBB loss preprocessing! This is an automated message to help with review readiness, and an Ultralytics engineer will assist you soon. Please review the checklist below:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Comment by Laughing-q** [comment]
tested yolo26n training with bs=128:
- first epoch from 5:58 to 5:35

tested both yolo26n and yolo26n-obb training on both this branch and main branch, results(metrics and losses) stay the same.

File: ultralytics/utils/loss.py [hunk 1]
```diff
@@ -366,14 +366,15 @@ def preprocess(self, targets: torch.Tensor, batch_size: int, scale_tensor: torch
...
             counts = counts.to(dtype=torch.int32)
             out = torch.zeros(batch_size, counts.max(), ne - 1, device=self.device)
-            for j in range(batch_size):
-                matches = i == j
-                if n := matches.sum():
-                    out[j, :n] = targets[matches, 1:]
+            offsets = torch.zeros(batch_size + 1, dtype=torch.long, device=self.device)
+            offsets.scatter_add_(0, batch_idx + 1, torch.ones_like(batch_idx))
+            offsets = offsets.cumsum(0)
+            within_idx = torch.arange(nl, device=self.device) - offsets[batch_idx]
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/loss.py:376]: ⚠️ **HIGH**: This indexing scheme only works if `targets` is already grouped by `batch_idx`. When image indices are interleaved (for example `[1, 0, 1]`), `within_idx = arange(...) - offsets[batch_idx]` produces negative or duplicated positions, so targets get written to the wrong slots or raise an indexing error. The previous per-image masking handled arbitrary target order correctly. The same issue exists in the mirrored OBB preprocess block below.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Good vectorization direction, but the new packing logic is not order-safe: it assumes targets are pre-sorted by image index, which the previous implementation did not require. That needs to be fixed before this change is safe to merge.

💬 Posted 1 inline comment

**Review by lmycross — APPROVED** [review]

**Comment by UltralyticsAssistant** [comment]
🎉 Wonderful to see this merged — huge thanks to @Laughing-q, with great contributions from @lmycross and @ahmet-f-gumustas!

This is a thoughtful and impactful optimization: speeding up `loss.py` target preprocessing through cleaner vectorized PyTorch indexing while preserving behavior is exactly the kind of work that strengthens Ultralytics from the inside out. As Henry Ford said, *“Coming together is a beginning; keeping together is progress; working together is success.”* This PR reflects that spirit perfectly — collaborative, precise, and meaningfully improving training efficiency for everyone.

Really appreciate the care behind this improvement and the value it brings to the YOLO community 🚀
